### PR TITLE
#3 created class for ai image generation and a demo

### PR DESCRIPTION
**Pull request by nathan-probert**
Built a class for ai image generation, runs 'out of the box' but may/will require further testing and experimentation with hyper parameters to achieve good, consistent results.

Currently supports image to image and text to image.

Also created an asynchronous queue such that the script will generate one image at a time (to save the GPU) but will run another thread that can check for input to allow continuous image generation. The example used is a simple one, but this could allow us to leave the script running and send images/prompts asynchronously, e.g even link it to a database or another script.

A lot of it was trial and error, so if there is anything to change off the top of your heads let me know while most of it is still fresh in my mind.

### CODE REVIEW

File: ImageHandler.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+        task = {'type': 'prompt', 'kwargs': kwargs}
+        self.queue.put(task)
+        print(f"\nEnqueued: {prompt}")
+    
+    # Continuously process the queue
+    def _process_queue(self):
+        while True:
+            task = self.queue.get()
+            if task is None:
```
Comment (zachale) [inline ImageHandler.py]: nit: could it be best practice for readability to put the positive case in the conditional?
```suggestion
            if task:
                print(f"\nDequeued: {task['kwargs']['prompt']}")
                self.process(task)
                self.queue.task_done()
            else:
                 break
```

File: ImageHandler.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+        return image
+    
+    # Function to add image to image task to queue
+    def enqueue_image_to_image(self, init_image: Union[PIL.Image.Image, str], prompt = 'make photograph', num_inference_steps=None, guidance_scale=None):
+        if isinstance(init_image, str):
+            init_image = self.get_image_from_url(init_image)
+
+        kwargs = {'image': init_image, 'prompt': prompt}
+
+        if num_inference_steps is not None:
```
Comment (zachale) [inline ImageHandler.py]: I think you could [simplify ](https://note.nkmk.me/en/python-none-usage) this if statement
```suggestion
        if num_inference_steps:
```

File: ImageHandler.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+            kwargs['num_inference_steps'] = num_inference_steps
+        if guidance_scale is not None:
```
Comment (zachale) [inline ImageHandler.py]: There are a handful of if statements just like this. Take a look at them and see if it's possible to optimize them!

**Review by zachale — CHANGES_REQUESTED** [review]
LGTM! 🚀 
Great First PR! I have a few minor comments, let me know if I am wrong.

**Comment by faisal-fawad** [comment]
Pretty good! Cool data structure and clean usage of kwargs :) Apart from Zach's suggestions I don't think there is anything really... I did notice that there are two possible types: "image" and "prompt"; instead of using a string here it may be better practice to use an enum (?) though it isn't too big of a deal

**Review by faisal-fawad — APPROVED** [review]

**Review by zachale — APPROVED** [review]

**Pull request merged**